gateway/types.go

```diff
@@ -92,6 +111,25 @@ type Checkpoint struct {
 	Sig  []byte
 }
 
+func NewCheckpoint(id sdk.SubnetID, epoch abi.ChainEpoch) *Checkpoint {
+	return &Checkpoint{
+		Data: CheckData{Source: id, Epoch: epoch},
+	}
+}
+
+func (c *Checkpoint) Cid() (cid.Cid, error) {
+	buf := new(bytes.Buffer)
+	if err := c.Data.MarshalCBOR(buf); err != nil {
+		return cid.Undef, err
+	}
+	h, err := mh.Sum(buf.Bytes(), abi.HashFunction, -1)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	return cid.NewCidV1(abi.CidBuilder.GetCodec(), h), nil
+}
+
 func (c *Checkpoint) CrossMsgMeta(from, to *sdk.SubnetID) (*CrossMsgMeta, bool) {
 	for i, m := range c.Data.CrossMsgs {
```

Comment:
Why are we sure that CrossMsgs field is not nil?

Reply:
We are not, Alex worked on that and I haven't reviewed those methods yet. Will fix this one now.